CoursesManager.MVVM/GlobalCache/GlobalCache.cs

```diff
@@ -40,32 +42,53 @@ public object Get(string key)
 
     public void Put(string key, object value, bool isPermanent)
     {
+        ArgumentNullException.ThrowIfNull(value, nameof(value));
+
+
         lock (_lock)
         {
-            if (_cacheMap.TryGetValue(key, out var existingNode))
+            if (_cacheMap.TryGetValue(key, out var existingNode) && existingNode != null)
             {
+                if (existingNode.Value.IsPermanent)
+                {
+                    throw new CantBeOverwrittenException($"The item with key '{key}' is permanent and cannot be overwritten.");
+                }
+
+                // Instead of creating a new item (if the item is a non-permanent item), we update the item to reduce memory usage and improve performance. 
                 existingNode.Value = new CacheItem(key, value, isPermanent);
                 _usageOrder.Remove(existingNode);
                 _usageOrder.AddFirst(existingNode);
+                return;
             }
-            else
-            {
-                // Check if cache is full and only full of permanent items
-                if (_cacheMap.Count >= _capacity && _permanentItemCount == _capacity)
-                {
-                    IncreaseCapacity();
-                }
 
-                // Create and add the new node to the cache
-                var newNode = new LinkedListNode<CacheItem>(new CacheItem(key, value, isPermanent));
-                _usageOrder.AddFirst(newNode);
-                _cacheMap[key] = newNode;
+            EnsureCapacity();
 
-                // The permanent item count will be updated in the CacheItem class itself
-            }
+
+            // Create and add the new node to the cache
+            var newNode = new LinkedListNode<CacheItem>(new CacheItem(key, value, isPermanent));
+            _usageOrder.AddFirst(newNode);
+            _cacheMap[key] = newNode;
+
+            // The permanent item count will be updated in the CacheItem class itself
+        }
```

Comment:
Waarom heb je hier eigenlijk een lock? Het is toch een concurrent dictionary? Kom ook vandaag pas achter maar er is schijnbaar een methode op de ConcurrentDictionary die heet UpdateOrAdd. Misschien beter om dat te gebruiken? Dan maak je echt gebruik van het Concurrent deel.